src/circuit/circuit_frame.rs

```diff
@@ -4639,75 +4630,123 @@ pub fn comparison_helper<F: LurkField, CS: ConstraintSystem<F>>(
     Ok((is_comparison_tag, comp_val, diff_is_negative))
 }
 
-// Enforce 0 <= num < 2ˆn.
-pub fn enforce_at_most_n_bits<F: LurkField, CS: ConstraintSystem<F>>(
-    mut cs: CS,
-    g: &GlobalAllocations<F>,
-    num: AllocatedNum<F>,
-    n: usize,
-) -> Result<(), SynthesisError> {
-    let num_bits = num.to_bits_le(&mut cs.namespace(|| "u64 remainder bit decomp"))?;
-    let v = num_bits[n..255].to_vec();
-    popcount(&mut cs.namespace(|| "add all MSBs"), &v, &g.false_num);
-    Ok(())
+// Lurk supported uint coercion
+#[derive(Copy, Clone)]
+pub enum UnsignedInt {
+    U32,
+    U64,
 }
 
-// Convert from num to u64.
-pub fn to_u64<F: LurkField, CS: ConstraintSystem<F>>(
+impl UnsignedInt {
+    pub fn num_bits(&self) -> u32 {
+        match self {
+            UnsignedInt::U32 => 32,
+            UnsignedInt::U64 => 64,
+        }
+    }
+}
+
+pub fn to_unsigned_integer_helper<F: LurkField, CS: ConstraintSystem<F>>(
     mut cs: CS,
     g: &GlobalAllocations<F>,
-    maybe_u64: &AllocatedNum<F>,
+    field_elem: &AllocatedNum<F>,
+    field_bn: BigUint,
+    field_elem_bits: &[Boolean],
+    size: UnsignedInt,
 ) -> Result<AllocatedNum<F>, SynthesisError> {
-    let p64_bn = BigUint::pow(&BigUint::from_u32(2).unwrap(), 64);
-    let v = match maybe_u64.get_value() {
-        Some(v) => v,
-        None => F::zero(),
-    };
-    let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
-
-    let (q_bn, r_bn) = field_bn.div_rem(&p64_bn);
+    let power_of_two_bn = BigUint::pow(&BigUint::from_u32(2).unwrap(), size.num_bits());
 
+    let (q_bn, r_bn) = field_bn.div_rem(&power_of_two_bn);
     let q_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "q"), q_bn)?;
     let r_num = allocate_unconstrained_bignum(&mut cs.namespace(|| "r"), r_bn)?;
 
-    // a = b.q + r
-    let product = mul(
-        &mut cs.namespace(|| "product(q,pow(2,64))"),
-        &q_num,
-        &g.power2_64_num,
+    // field element = pow(2, size).q + r
+    match size {
+        UnsignedInt::U32 => linear(
+            &mut cs,
+            || "product(q,pow(2,32)) + r",
+            &q_num,
+            &g.power2_32_num,
+            &r_num,
+            field_elem,
+        ),
+        UnsignedInt::U64 => linear(
+            &mut cs,
+            || "product(q,pow(2,64)) + r",
+            &q_num,
+            &g.power2_64_num,
+            &r_num,
+            field_elem,
+        ),
+    };
```

Comment:
It would be clearer to just call `linear` once, after having assigned `g.power2_32_num` or `g.power2_64_num` to a variable.

Reply:
done